airbyte-integrations/bases/base-python/base_python/sdk/streams/http.py

```diff
@@ -0,0 +1,251 @@
+
+        E.g: you might want to define query parameters for paging if next_page_token is not None.
+        """
+        return {}
+
+    def get_request_headers(
+            self,
+            request_configuration: Optional[Dict] = None,
+            stream_state: Dict[str, Any] = {},
+            next_page_token: Dict[str, Any] = None,
+            parent_stream_record: Dict = None
+    ) -> Dict[str, Any]:
+        """
+        Override to return any non-auth headers. Authentication headers will overwrite any overlapping headers returned from this method.
+        """
+        return {}
+
+    def get_request_body_json(
+            self,
+            request_configuration: Optional[Dict] = None,
+            stream_state: Dict[str, Any] = {},
+            next_page_token: Dict[str, Any] = None,
+            parent_stream_record: Dict = None
+    ) -> Optional[Dict]:
+        """
+        TODO make this possible to do for non-JSON APIs
+        Override when creating POST requests to populate the body of the request with a JSON payload.
+        """
+        return None
+
+    def decode_response(self, response: requests.Response) -> Dict:
+        """
+        Decodes the response object from its raw form returned from the API to a Dict. By default this parses a JSON.
+        Override to parse responses differently e.g as XML or custom binary format.
+
+        :param response: The response object returned from the API endpoint.
+        :return: The response object decoded into a Dict. For example, using response.json() or response.text() wrapped in a Dict etc...
+        """
+        return response.json()
+
+    def parse_response(self, response: Dict) -> Iterable[Dict]:
+        """
+        Parses the raw response object into a list of records. This is useful in cases where a response contains more than one record e.g: a page
+        of 100 responses. This is also useful if you want to transform the incoming data e.g: filter some records or change some fields.
+
+        By default, this returns an iterable containing the input. Override to parse differently.
+        :param response:
+        :return: An iterable containing the parsed response
+        """
+        yield [response]
+
+    def should_backoff(self, response: requests.Response) -> bool:
+        """
+        Override to set different conditions for backoff based on the response from the server.
+
+        By default, back off on the following HTTP response statuses:
+         - 429 (Too Many Requests) indicating rate limiting
+         - 500s to handle transient server errors
+
+        Unexpected but transient exceptions (connection timeout, DNS resolution failed, etc..) are retried by default.
+        """
+        return response.status_code == 429 or 500 <= response.status_code < 600
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
```

Comment:
Maybe this should specify behavior for a backoff instead of just allowing time? Like attempt number, response, and then allow calculating time? Then someone could do exp. backoff scaled their own way for example.

Reply:
Also it feels like this should be a `Duration` equivalent or milliseconds instead of seconds-level.

Reply:
Discussed in person. Sounds like this will move towards custom decorators/handlers instead of these function overrides.

Reply:
fwiw this might be a lot cleaner if we can get this merged: https://github.com/litl/backoff/pull/122